File: backend/app/services/normalizers.py
"""Response normalizers — raw API dicts → unified Pydantic models.

Each `normalize_*` function accepts a *validated* raw dict and returns
the appropriate unified schema model.  Validation (field existence checks)
happens in validators.py; normalization (field mapping, type coercion,
defaults) happens here.
"""
from __future__ import annotations

import logging
from datetime import datetime
from typing import Any

from app.services.schemas import (
    EconomyPhase,
    MarketOffer,
    MarketSnapshot,
    ResourceEntry,
    RetailInfo,
    SimcoResource,
)

log = logging.getLogger(__name__)

_PHASE_NAMES = {0: "Stable", 1: "Boom", 2: "Recession", 3: "Recovery"}
_PHASE_MULTIPLIERS = {0: 1.0, 1: 1.25, 2: 0.80, 3: 0.95}


# ---------------------------------------------------------------------------
# Market offers
# ---------------------------------------------------------------------------

def normalize_market_offer(raw: dict, product_id: int, realm: int) -> MarketOffer:
    """Map a single raw offer dict to MarketOffer.

    SimCompanies API uses camelCase; some community scrapers use snake_case.
    We handle both.
    """
    return MarketOffer(
        offer_id   = int(raw.get("id") or raw.get("offer_id", 0)),
        product_id = int(raw.get("kind", raw.get("product_id", product_id))),
        realm      = int(raw.get("realm", realm)),
        quality    = int(raw.get("quality", 1)),
        price      = float(raw.get("price") or raw.get("unitCost", 0)),
        quantity   = int(raw.get("quantity") or raw.get("amount", 0)),
        seller_id  = raw.get("company") or raw.get("seller_id"),
    )


def normalize_market_snapshot(
    offers: list[MarketOffer],
    product_id: int,
    realm: int,
    quality: int = 1,
) -> MarketSnapshot:
    """Aggregate a list of offers into one MarketSnapshot."""
    filtered = [o for o in offers if o.quality == quality]

    if not filtered:
        return MarketSnapshot(
            product_id=product_id,
            realm=realm,
            quality=quality,
            lowest_ask=0.0,
            highest_ask=0.0,
        )

    prices    = [o.price for o in filtered]
    quantities = [o.quantity for o in filtered]
    total_qty  = sum(quantities)

    vwap = (
        sum(p * q for p, q in zip(prices, quantities)) / total_qty
        if total_qty else 0.0
    )

    if len(prices) > 1:
        avg   = sum(prices) / len(prices)
        variance  = sum((p - avg) ** 2 for p in prices) / len(prices)
        volatility = (variance ** 0.5) / avg if avg else 0.0
    else:
        volatility = 0.0

    return MarketSnapshot(
        product_id      = product_id,
        realm           = realm,
        quality         = quality,
        lowest_ask      = min(prices),
        highest_ask     = max(prices),
        vwap            = round(vwap, 6),
        total_supply    = total_qty,
        offer_count     = len(filtered),
        demand_score    = min(1.0, total_qty / 1000),  # heuristic; enriched later
        price_volatility = round(volatility, 6),
        source          = "simcompanies",
    )


# ---------------------------------------------------------------------------
# Encyclopedia / resource entries
# ---------------------------------------------------------------------------

# Category mapping: SimCompanies numeric kind → string category
_CATEGORY_MAP: dict[int, str] = {
    0: "agriculture",   1: "agriculture",  2: "agriculture",
    10: "mining",       11: "mining",
    20: "manufacturing",21: "manufacturing",
    30: "electronics",  31: "electronics",
    40: "chemicals",    41: "chemicals",
    50: "automotive",   51: "automotive",
    60: "aerospace",    61: "aerospace",
    70: "food",         71: "food",
    80: "retail",       81: "retail",
    90: "research",
    100: "real_estate",
}


def normalize_resource_entry(raw: dict, realm: int) -> ResourceEntry:
    """Map a raw encyclopedia resource dict to ResourceEntry."""
    kind_id = int(raw.get("id") or raw.get("kind", 0))
    category_raw = raw.get("category", "")
    if isinstance(category_raw, int):
        category = _CATEGORY_MAP.get(category_raw, "other")
    else:
        category = str(category_raw).lower() if category_raw else _CATEGORY_MAP.get(kind_id // 10 * 10, "other")

    ingredients_raw = raw.get("ingredients") or raw.get("recipe") or []
    if isinstance(ingredients_raw, dict):
        ingredients = [{"product_id": k, "quantity": v} for k, v in ingredients_raw.items()]
    else:
        ingredients = ingredients_raw

    return ResourceEntry(
        sim_id           = kind_id,
        realm            = realm,
        key              = raw.get("db_letter") or raw.get("key") or raw.get("name", ""),
        name             = raw.get("name", ""),
        category         = category,
        retail_price     = float(raw.get("retailPrice") or raw.get("retail_price") or 0),
        transport_cost   = float(raw.get("transportCost") or raw.get("transport_cost") or 0),
        units_per_run    = int(raw.get("unitsPerRun") or raw.get("units_per_run") or 1),
        production_time  = int(raw.get("productionTime") or raw.get("production_time") or 0),
        is_raw_material  = bool(raw.get("isRawMaterial") or raw.get("is_raw_material") or False),
        is_tradeable     = bool(raw.get("isTradeable") if "isTradeable" in raw else raw.get("is_tradeable", True)),
        ingredients      = ingredients,
    )


# ---------------------------------------------------------------------------
# Retail info
# ---------------------------------------------------------------------------

def normalize_retail_info(raw: dict, realm: int) -> RetailInfo:
    product_id = int(raw.get("id") or raw.get("resource") or raw.get("product_id", 0))
    quality    = int(raw.get("quality", 1))
    return RetailInfo(
        product_id          = product_id,
        realm               = realm,
        quality             = quality,
        daily_demand        = float(raw.get("dailyDemand") or raw.get("daily_demand") or 0),
        max_retail_price    = float(raw.get("maxRetailPrice") or raw.get("max_retail_price") or 0),
        saturation_level    = float(raw.get("saturationLevel") or raw.get("saturation_level") or 0),
        quality_multiplier  = float(raw.get("qualityMultiplier") or raw.get("quality_multiplier") or 1.0),
    )


# ---------------------------------------------------------------------------
# Economy phase
# ---------------------------------------------------------------------------

def normalize_economy_phase(raw: dict, realm: int) -> EconomyPhase:
    data = raw.get("data", raw)
    code = int(data.get("phase") or data.get("phase_code") or 0)
    name = data.get("name") or data.get("phase_name") or _PHASE_NAMES.get(code, "Unknown")
    mult = float(data.get("multiplier") or _PHASE_MULTIPLIERS.get(code, 1.0))

    started_raw = data.get("startedAt") or data.get("started_at")
    started_at: datetime | None = None
    if started_raw:
        try:
            started_at = datetime.fromisoformat(str(started_raw).replace("Z", "+00:00"))
        except ValueError:
            pass

    return EconomyPhase(
        realm      = realm,
        phase_code = code,
        phase_name = name,
        multiplier = mult,
        started_at = started_at,
        source     = data.get("source", "simcotools"),
    )


# ---------------------------------------------------------------------------
# SimcoTools enriched resources
# ---------------------------------------------------------------------------

def normalize_simco_resource(raw: dict, realm: int) -> SimcoResource:
    sim_id   = int(raw.get("id") or raw.get("sim_id") or 0)
    updated_raw = raw.get("updatedAt") or raw.get("updated_at")
    updated_at: datetime | None = None
    if updated_raw:
        try:
            updated_at = datetime.fromisoformat(str(updated_raw).replace("Z", "+00:00"))
        except ValueError:
            pass

    return SimcoResource(
        sim_id           = sim_id,
        realm            = realm,
        name             = raw.get("name", ""),
        category         = str(raw.get("category", "other")).lower(),
        avg_price        = float(raw.get("avgPrice") or raw.get("avg_price") or 0),
        min_price        = float(raw.get("minPrice") or raw.get("min_price") or 0),
        max_price        = float(raw.get("maxPrice") or raw.get("max_price") or 0),
        price_trend      = float(raw.get("priceTrend") or raw.get("price_trend") or 0),
        total_supply     = int(raw.get("totalSupply") or raw.get("total_supply") or 0),
        demand_index     = float(raw.get("demandIndex") or raw.get("demand_index") or 0),
        volatility_index = float(raw.get("volatilityIndex") or raw.get("volatility_index") or 0),
        updated_at       = updated_at,
    )

File: extension/src/services/normalizers.js
/**
 * normalizers.js — Raw API response → Unified Internal Schema
 *
 * Each function accepts the raw JSON from a specific provider/endpoint
 * and returns a shape defined in schema.js.
 *
 * Normalizers must:
 *  - Never throw — return null for invalid records so callers can filter
 *  - Coerce types (string → number, missing → default)
 *  - Sanitise values (negative prices → 0, etc.)
 *
 * @module services/normalizers
 */

'use strict';

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

const safeNum   = (v, fallback = 0) => { const n = parseFloat(v); return isFinite(n) ? n : fallback; };
const safeInt   = (v, fallback = 0) => { const n = parseInt(v, 10); return isFinite(n) ? n : fallback; };
const clamp     = (v, min, max) => Math.max(min, Math.min(max, v));
const isoNow    = () => new Date().toISOString();
const toISO     = v  => { try { return new Date(v).toISOString(); } catch { return isoNow(); } };

// ---------------------------------------------------------------------------
// SimCompanies — market offers  (/api/v2/market/{itemId})
// ---------------------------------------------------------------------------

/**
 * Normalise a single raw market offer from SimCompanies.
 *
 * @param {object} raw    - One element from the API array
 * @param {number} realm
 * @returns {import('./schema.js').MarketOffer|null}
 */
export function normalizeSimcoOffer(raw, realm) {
  if (!raw || typeof raw !== 'object') return null;
  const price = safeNum(raw.price ?? raw.Price);
  const qty   = safeInt(raw.quantity ?? raw.Quantity ?? raw.qty);
  if (price <= 0 || qty <= 0) return null;

  return {
    productId: safeInt(raw.resource ?? raw.resourceId ?? raw.id),
    realm,
    price,
    quantity: qty,
    quality:  clamp(safeInt(raw.quality ?? raw.Quality), 0, 3),
    sellerId: String(raw.seller?.id ?? raw.sellerId ?? raw.company ?? ''),
    listedAt: toISO(raw.posted ?? raw.listedAt ?? raw.created_at),
  };
}

/**
 * Normalise a full SimCompanies market response into a MarketSnapshot.
 *
 * @param {any}    raw
 * @param {number} productId
 * @param {number} realm
 * @returns {import('./schema.js').MarketSnapshot|null}
 */
export function normalizeSimcoMarket(raw, productId, realm) {
  // SimCo returns an array of offers directly
  const offerList = Array.isArray(raw) ? raw : (raw?.offers ?? raw?.data ?? []);
  const offers = offerList
    .map(o => normalizeSimcoOffer(o, realm))
    .filter(Boolean)
    .sort((a, b) => a.price - b.price);

  if (offers.length === 0) {
    return {
      productId, realm,
      lowestAsk: 0, highestAsk: 0, vwap: 0,
      totalSupply: 0, offerCount: 0,
      offers: [],
      fetchedAt: isoNow(),
    };
  }

  const totalQty   = offers.reduce((s, o) => s + o.quantity, 0);
  const totalValue = offers.reduce((s, o) => s + o.price * o.quantity, 0);

  return {
    productId,
    realm,
    lowestAsk:   offers[0].price,
    highestAsk:  offers[offers.length - 1].price,
    vwap:        totalQty > 0 ? totalValue / totalQty : 0,
    totalSupply: totalQty,
    offerCount:  offers.length,
    offers,
    fetchedAt:   isoNow(),
  };
}

// ---------------------------------------------------------------------------
// SimCompanies — encyclopedia resource  (/api/v4/pt/{realm}/encyclopedia/resources/)
// ---------------------------------------------------------------------------

/**
 * @param {object} raw
 * @param {number} realm
 * @returns {import('./schema.js').ResourceInfo|null}
 */
export function normalizeSimcoResource(raw, realm) {
  if (!raw || typeof raw !== 'object') return null;
  const id = safeInt(raw.id ?? raw.dbLetter ?? raw.resourceId);
  if (!id) return null;

  // inputs can be under `inputs`, `recipe`, or `resources`
  const rawInputs = raw.inputs ?? raw.recipe ?? raw.resources ?? [];
  const inputs = Array.isArray(rawInputs)
    ? rawInputs.map(i => ({
        resourceId: safeInt(i.resource ?? i.resourceId ?? i.id),
        quantity:   safeNum(i.quantity ?? i.amount ?? i.qty),
      })).filter(i => i.resourceId > 0)
    : [];

  return {
    id,
    realm,
    key:            String(raw.key ?? raw.name_key ?? '').toLowerCase(),
    name:           String(raw.name ?? raw.label ?? ''),
    category:       String(raw.kind ?? raw.category ?? '').toLowerCase(),
    retailPrice:    safeNum(raw.retailPrice ?? raw.retail_price ?? raw.retail),
    transportCost:  safeNum(raw.transportationCost ?? raw.transport_cost ?? raw.transport),
    unitsPerRun:    safeInt(raw.unitsPerRun ?? raw.units_per_run ?? raw.output ?? 1),
    productionTime: safeInt(raw.productionTime ?? raw.production_time ?? raw.time ?? 0),
    isRawMaterial:  Boolean(raw.rawMaterial ?? raw.raw_material ?? raw.natural_resource),
    inputs,
  };
}

// ---------------------------------------------------------------------------
// SimCompanies — economy phase  (phase endpoint)
// ---------------------------------------------------------------------------

/**
 * @param {object} raw
 * @param {number} realm
 * @returns {import('./schema.js').EconomyPhase|null}
 */
export function normalizeSimcoPhase(raw, realm) {
  if (!raw || typeof raw !== 'object') return null;

  const PHASE_NAMES = { 0: 'Stable', 1: 'Boom', 2: 'Recession', 3: 'Recovery' };
  const MULTIPLIERS = { 0: 1.0,     1: 1.25,  2: 0.80,        3: 0.95 };

  const code = safeInt(raw.phase ?? raw.phaseCode ?? raw.state ?? 0);

  return {
    realm,
    phaseCode:  code,
    phaseName:  raw.phaseName ?? raw.name ?? PHASE_NAMES[code] ?? 'Unknown',
    multiplier: safeNum(raw.multiplier ?? MULTIPLIERS[code] ?? 1.0),
    fetchedAt:  isoNow(),
  };
}

// ---------------------------------------------------------------------------
// SimCompanies — retail info  (/api/v4/{realm}/resources-retail-info/)
// ---------------------------------------------------------------------------

/**
 * @param {object} raw    - one element from the retail-info array
 * @param {number} realm
 * @returns {import('./schema.js').RetailInfo|null}
 */
export function normalizeSimcoRetailInfo(raw, realm) {
  if (!raw || typeof raw !== 'object') return null;
  const productId = safeInt(raw.resource ?? raw.id ?? raw.productId);
  if (!productId) return null;

  return {
    productId,
    realm,
    retailPrice:    safeNum(raw.retailPrice ?? raw.retail_price ?? raw.price),
    maxRetailPrice: safeNum(raw.maxRetailPrice ?? raw.max_retail_price ?? raw.maxPrice),
    demand:         clamp(safeNum(raw.demand ?? raw.demandFactor ?? 0), 0, 1),
    fetchedAt:      isoNow(),
  };
}

// ---------------------------------------------------------------------------
// SimcoTools — resources  (https://simcotools.app/api/v3/resources)
// ---------------------------------------------------------------------------

/**
 * @param {object} raw
 * @returns {import('./schema.js').SimcoToolsResource|null}
 */
export function normalizeSimcoToolsResource(raw) {
  if (!raw || typeof raw !== 'object') return null;
  const id = safeInt(raw.id ?? raw.resourceId);
  if (!id) return null;

  return {
    id,
    key:          String(raw.key ?? raw.slug ?? '').toLowerCase(),
    name:         String(raw.name ?? ''),
    category:     String(raw.kind ?? raw.category ?? '').toLowerCase(),
    avgPrice24h:  raw.avgPrice24h != null ? safeNum(raw.avgPrice24h) : null,
    avgPrice7d:   raw.avgPrice7d  != null ? safeNum(raw.avgPrice7d)  : null,
    avgVolume24h: raw.avgVolume24h != null ? safeNum(raw.avgVolume24h) : null,
    volatility:   safeNum(raw.volatility ?? raw.priceVolatility ?? 0),
    fetchedAt:    isoNow(),
  };
}

// ---------------------------------------------------------------------------
// SimcoTools — phases  (https://api.simcotools.com/v1/realms/{realm}/phases)
// ---------------------------------------------------------------------------

/**
 * @param {object} raw
 * @param {number} realm
 * @returns {import('./schema.js').EconomyPhase|null}
 */
export function normalizeSimcoToolsPhase(raw, realm) {
  if (!raw || typeof raw !== 'object') return null;

  const PHASE_MAP = {
    stable:   { code: 0, multiplier: 1.00 },
    boom:     { code: 1, multiplier: 1.25 },
    recession:{ code: 2, multiplier: 0.80 },
    recovery: { code: 3, multiplier: 0.95 },
  };

  const nameRaw = String(raw.phase ?? raw.name ?? raw.state ?? 'stable').toLowerCase();
  const mapped  = PHASE_MAP[nameRaw] ?? PHASE_MAP.stable;

  return {
    realm,
    phaseCode:  mapped.code,
    phaseName:  nameRaw.charAt(0).toUpperCase() + nameRaw.slice(1),
    multiplier: safeNum(raw.multiplier ?? mapped.multiplier),
    fetchedAt:  isoNow(),
  };
}
